netutil/netlink.go

```diff
@@ -0,0 +1,172 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package netutil
+
+import (
+	"fmt"
+	"net"
+	"syscall"
+)
+
+const (
+	// see /usr/include/linux/rtnetlink.h
+	RTMGRP_IPV4_ROUTE = 0x40
+	RTMGRP_IPV6_ROUTE = 0x400
+
+	RTM_NEWROUTE = 24
+	RTM_DELROUTE = 25
+)
+
+// openNetlinkFd is used in tests to mock a netlink socket
+var openNetlinkFd = openNetlinkFdImpl
+
+func openNetlinkFdImpl() (fd int, err error) {
+	fd, err = syscall.Socket(syscall.AF_NETLINK, syscall.SOCK_RAW, syscall.NETLINK_ROUTE)
+	if err != nil {
+		return -1, err
+	}
+	addr := &syscall.SockaddrNetlink{
+		Family: syscall.AF_NETLINK,
+		Groups: RTMGRP_IPV4_ROUTE | RTMGRP_IPV6_ROUTE,
+	}
+	if err := syscall.Bind(fd, addr); err != nil {
+		return -1, err
+	}
+	return fd, nil
+}
+
+// RoutesMonitor monitors the network information and provides callbacks
+// when a default gateway is added or removed
+type RoutesMonitor struct {
+	netlinkFd int
+
+	netlinkErrors chan error
+	netlinkQuitCh chan struct{}
+
+	defaultGwAdded   func(string)
+	defaultGwRemoved func(string)
+}
+
+func NewRoutesMonitor(defaultGwAdded, defaultGwRemoved func(string)) *RoutesMonitor {
+	m := &RoutesMonitor{
+		netlinkFd: -1,
+
+		netlinkErrors: make(chan error),
+
+		defaultGwAdded:   defaultGwAdded,
+		defaultGwRemoved: defaultGwRemoved,
+	}
+	return m
+}
+
+func (m *RoutesMonitor) Connect() error {
+	if m.netlinkFd != -1 {
+		return fmt.Errorf("cannot connect: already connected")
+	}
+
+	fd, err := openNetlinkFd()
+	if err != nil {
+		return err
+	}
+	m.netlinkFd = fd
+	return nil
+}
+
+func (m *RoutesMonitor) Stop() {
+	close(m.netlinkQuitCh)
```

Comment:
our stop usually waits for the thing to be stopped, I kind of recommend using a Tomb it might the fastest path to getting what we expect

Reply:
@mvo5 maybe I was confusing but afaict the syscall.Close here was fine, it was the assignment to netlinkFD that was problematic

Reply:
Thank you! Stoping can take an arbitrary amount of time AFACT. The issue is that the "monitor()" go routine can be in "syscall.Read(m.netlinkFd)". This call is blocking so only after some data is received there the montior() can react to the closing of the netlinkQuitCh. Here is a gist that demos the issue https://gist.github.com/mvo5/fa14638782fe30949998096d7b3c6314. But maybe I'm just missing something here.

Reply:
what about reacting to closing the FD? we can do that

Reply:
Ok, I made it work but definitely it's ugly (but not a lot of extra code), because this kind of socket atm cannot be either made into a non-blocking os.File nor any kind of netpoll/net supported connection as you remarked.  I had to use sycall.Select and a os.Pipe as well. I need to clean up what I have.